File: guests/rust/background.rs
//! The background scene (tools/scenes/background.steps): the worker parks
//! for a CLICK, so a binding that used the app thread DEADLOCKS here.

use std::sync::mpsc;
use std::sync::{Arc, Mutex};

pub(crate) fn app(ctx: kaya::AppCtx) {
    #[derive(Clone, Copy)]
    enum Msg {
        Start,
        Ping,
        Release,
        Nest,
    }

    let msgs = kaya::Messages::<Msg>::new();
    let (status, alive, nested) = ctx.apply(|tx| {
        tx.window(kaya::DEFAULT_WINDOW).title("background");
        let status = tx.signal("idle");
        let alive = tx.signal("-");
        let nested = tx.signal("-");
        let root = tx
            .column(|tx| {
                tx.label(status).a11y_id("status"); // label#0
                tx.label(alive).a11y_id("alive"); // label#1
                // Addressed by id: an index read passes for an empty arm.
                tx.label(nested).a11y_id("nested"); // label#2
                let start = tx.button("start").id(); // button#0
                msgs.on_click(start, Msg::Start);
                let ping = tx.button("ping").id(); // button#1
                msgs.on_click(ping, Msg::Ping);
                let release = tx.button("release").id(); // button#2
                msgs.on_click(release, Msg::Release);
                let nest = tx.button("nest").id(); // button#3
                msgs.on_click(nest, Msg::Nest);
            })
            .id();
        tx.mount(root);
        (status, alive, nested)
    });

    // The send must NOT wait for the receiver — mpsc's does not.
    let (release_tx, release_rx) = mpsc::channel::<()>();
    let mut release_tx = Some(release_tx);
    let mut release_rx = Some(release_rx);

    while let Some(msg) = msgs.next(&ctx) {
        match msg {
            Msg::Start => {
                let poster = ctx.poster();
                let rx = release_rx.take().expect("start is clicked once");
                let acc = Arc::new(Mutex::new(String::new()));
                std::thread::Builder::new()
                    .name("background-worker".into())
                    .spawn(move || {
                        // Parks: on the app thread that click could never
                        // be processed.
                        let _ = rx.recv();
                        for step in ["1", "2", "3"] {
                            let acc = Arc::clone(&acc);
                            poster.post(move |tx| {
                                let mut seen = acc.lock().unwrap();
                                seen.push_str(step);
                                tx.write(status, seen.clone());
                            });
                        }
                    })
                    .expect("failed to spawn the background worker");
                ctx.apply(|tx| tx.write(status, "working"));
            }
            Msg::Ping => ctx.apply(|tx| tx.write(alive, "alive")),
            Msg::Release => {
                if let Some(tx) = release_tx.take() {
                    let _ = tx.send(());
                }
            }
            // A post from INSIDE a transaction QUEUES for after.
            Msg::Nest => {
                let poster = ctx.poster();
                let seq = Arc::new(Mutex::new(String::new()));
                let inner = Arc::clone(&seq);
                ctx.apply(|tx| {
                    seq.lock().unwrap().push('a');
                    poster.post(move |tx| {
                        let mut s = inner.lock().unwrap();
                        s.push('b');
                        tx.write(nested, s.clone());
                    });
                    seq.lock().unwrap().push('c');
                    let now = seq.lock().unwrap().clone();
                    tx.write(nested, now);
                });
            }
        }
    }
}

fn main() {
    kaya::run(app)
}

File: guests/haskell/background.hs
{-# LANGUAGE OverloadedStrings #-}

-- The background scene, Haskell port — guests/rust/background.rs,
-- tools/scenes/background.steps.

import Control.Concurrent (forkIO, newEmptyMVar, takeMVar, tryPutMVar)
import Data.IORef (modifyIORef', newIORef, readIORef)
import Data.Text (Text)
import qualified Data.Text as T
import KayaApp

main :: IO ()
main = kayaMain $ \app -> do
  released <- newEmptyMVar
  postedRef <- newIORef ""
  nestedRef <- newIORef ""

  _ <- buildTx app $ do
    window primary [WTitle "background"]
    status <- signal (T.pack "idle")
    alive <- signal (T.pack "-")
    detail <- signal (T.pack "-")

    root <-
      column
        []
        [ labelBound status [A11yId ("status" :: Text)], -- label#0
          labelBound alive [A11yId ("alive" :: Text)], -- label#1
          labelBound detail [A11yId ("nested" :: Text)], -- label#2
          buttonOn
            "start" -- button#0
            ( do
                _ <- forkIO $ do
                  -- Parks until the scene clicks release.
                  takeMVar released
                  mapM_
                    ( \step -> post app $ do
                        modifyIORef' postedRef (++ step)
                        seen <- readIORef postedRef
                        buildTx app (writeSignal status (T.pack seen))
                    )
                    ["1", "2", "3"]
                buildTx app (writeSignal status (T.pack "working"))
            )
            [],
          buttonOn "ping" (buildTx app (writeSignal alive (T.pack "alive"))) [], -- button#1
          -- tryPutMVar, NOT putMVar: putMVar BLOCKS on a full MVar, so a
          -- second release click would wedge the APP THREAD (docs/deferred.md,
          -- the blocking-release entry).
          buttonOn "release" (() <$ tryPutMVar released ()) [], -- button#2
          buttonOn
            "nest" -- button#3
            ( do
                modifyIORef' nestedRef (++ "a")
                post app $ do
                  modifyIORef' nestedRef (++ "b")
                  seen <- readIORef nestedRef
                  buildTx app (writeSignal detail (T.pack seen))
                modifyIORef' nestedRef (++ "c")
                seen <- readIORef nestedRef
                buildTx app (writeSignal detail (T.pack seen))
            )
            []
        ]
    mount root
    return ()
  return ()
